src/Workspaces/Remote/Core/BrokeredServiceConnection.cs

```diff
@@ -143,19 +144,14 @@ public override async ValueTask<Optional<TResult>> TryInvokeAsync<TResult>(Solut
 
         internal static async ValueTask<TResult> InvokeStreamingServiceAsync<TResult>(
             TService service,
-            Func<TService, Stream, CancellationToken, ValueTask> invocation,
-            Func<Stream, CancellationToken, ValueTask<TResult>> reader,
+            Func<TService, PipeWriter, CancellationToken, ValueTask> invocation,
+            Func<PipeReader, CancellationToken, ValueTask<TResult>> reader,
             CancellationToken cancellationToken)
         {
-            // The reader should close the client stream, the writer will close the server stream.
-            // See https://github.com/microsoft/vs-streamjsonrpc/blob/master/doc/oob_streams.md
-            var (clientStream, serverStream) = FullDuplexStream.CreatePair();
-
-            // Create new tasks that both start executing, rather than invoking the delegates directly.
-            // If the reader started synchronously reading before the writer task started it would be blocking, and vice versa
-            // if the writer synchronously filled the buffer before the reader task started it would also be blocking.
-            var writerTask = Task.Run(async () => await invocation(service, serverStream, cancellationToken).ConfigureAwait(false), cancellationToken);
-            var readerTask = Task.Run(async () => await reader(clientStream, cancellationToken).ConfigureAwait(false), cancellationToken);
+            var pipe = new Pipe();
```

Comment:
i know nothing about pipes.  do they need to be 'using'ed/disposed/flushed?  or can you literally just instantiate and leave alone like this?

Reply:
The pipe itself doesn't. Each pipe has PipeReader and PipeWriter. They need to be completed, which is what `invocation` and `reader` operations are expected to do.